### PR TITLE
[RISCV] Fix mgather -> riscv.masked.strided.load combine not extending indices

### LINKED ISSUE #82430: RISCV vector zvl256b miscompile at -O2
Testcase:
```c
int printf(const char *, ...);
int c;
int d[10][10];
char e, f;
int main() {
  for (f = 0; f < 8; f += 1) {
    --c >> d[0][4] && (e = 0);
    d[0][4] = d[5][0];
  }
  printf("%X\n", d[0][4]);
}
```

Commands:
```bash
> /scratch/tc-testing/tc-feb-20-llvm/build/bin/clang -march=rv64gcv_zvl256b -O2 red.c -o red.out
> /scratch/tc-testing/tc-feb-20-llvm/build/bin/qemu-riscv64 red.out
AB2B8CB8
> /scratch/tc-testing/tc-feb-20-llvm/build/bin/clang -march=rv64gcv_zvl256b -O2 red.c -o red.out -fuse-ld=lld
> /scratch/tc-testing/tc-feb-20-llvm/build/bin/qemu-riscv64 red.out
555599E0
> /scratch/tc-testing/tc-feb-20-llvm/build/bin/clang red.c -o red.out
> /scratch/tc-testing/tc-feb-20-llvm/build/bin/qemu-riscv64 red.out
0
```

Bugpoint, ASM, red.c: [c-asm-bc.zip](https://github.com/llvm/llvm-project/files/14351798/c-asm-bc.zip)

`--opt-bisect-limit` points to `SLPVectorizerPass`

Discovered/tested using 7f3980a7b2c9f95ab3b106a94fe6e63158155b0b (not bisected)
Found using fuzzer

### PR DESCRIPTION
**Pull request by lukel97**
This fixes the miscompile reported in #82430 by telling isSimpleVIDSequence to sign extend to XLen instead of the width of the indices, since the "sequence" of indices generated by a strided load will be at XLen.

This was the simplest way I could think of getting isSimpleVIDSequence to treat the indexes as if they were zero extended to XLenVT.

Another way we could do this is by refactoring out the "get constant integers" part from isSimpleVIDSequence and handle them as APInts so we can separately zero extend it.

Fixes #82430

### CODE REVIEW

File: llvm/test/CodeGen/RISCV/rvv/fixed-vectors-masked-gather.ll [hunk 1]
```diff
@@ -15086,22 +15086,18 @@ define <32 x i64> @mgather_strided_split(ptr %base) {
...
-; FIXME: This is a miscompile triggered by the mgather ->
-; riscv.masked.strided.load combine. In order for it to trigger we need either a
-; strided gather that RISCVGatherScatterLowering doesn't pick up, or a new
-; strided gather generated by the widening sew combine.
 define <4 x i32> @masked_gather_widen_sew_negative_stride(ptr %base) {
 ; RV32V-LABEL: masked_gather_widen_sew_negative_stride:
 ; RV32V:       # %bb.0:
-; RV32V-NEXT:    addi a0, a0, -128
+; RV32V-NEXT:    addi a0, a0, 128
 ; RV32V-NEXT:    li a1, -128
```
Comment (topperc) [inline llvm/test/CodeGen/RISCV/rvv/fixed-vectors-masked-gather.ll:15093]: I think I would like to see something closer to the original test case where we also calculated the wrong stride. Here the stride was correct.

Comment (lukel97) [reply llvm/test/CodeGen/RISCV/rvv/fixed-vectors-masked-gather.ll]: I've updated the test in 11d115d0569b212dfeb7fe6485be48070e068e19 to make the stride larger than 128

**Review by topperc — APPROVED** [review]
LGTM

**Review by wangpc-pp — COMMENTED** [review]
LGTM.

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 2]
```diff
@@ -3261,14 +3262,14 @@ static std::optional<VIDSequence> isSimpleVIDSequence(SDValue Op) {
       if (!isa<ConstantSDNode>(Op.getOperand(Idx)))
         return std::nullopt;
       Val = Op.getConstantOperandVal(Idx) &
-            maskTrailingOnes<uint64_t>(EltSizeInBits);
+            maskTrailingOnes<uint64_t>(Op.getScalarValueSizeInBits());
```
[the remark this answers is not in the document]
Comment (lukel97) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp:3265]: I think we can also reduce the number of usages by only creating the list of integers once. I'll try and create a PR for this.

[the remark this answers is not in the document]
Comment (lukel97) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp:3265]: I've opened https://github.com/llvm/llvm-project/pull/82590, hopefully it removes some duplication

**Pull request merged**